api/samples/memval_simple.c

```diff
@@ -0,0 +1,434 @@
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Code Manipulation API Sample:
+ * memval_simple.c
+ *
+ * Records and dumps app write addresses, and their corresponding written values.
+ *
+ * (1) It fills two per-thread-buffers with inlined instrumentation.
+ * (2) Once the buffer has been filled up, a fault handler will redirect execution
+ *     to our trace buffer handler, where we dump the memrefs to disk.
+ *
+ * This sample illustrates
+ * - inserting instrumentation after the current instruction to read the value
+ *   written by it.
+ * - the use of drutil_expand_rep_string() to expand string loops to obtain
+ *   every memory reference,
+ * - the use of drutil_opnd_mem_size_in_bytes() to obtain the size of OP_enter
+ *   memory references,
+ * - the use of drutil_insert_get_mem_addr() to insert instructions to compute
+ *   the address of each memory reference.
+ * - the use of the drx_buf extension to fill buffers in a platform-independent
+ *   manner
+ *
+ * This client is a simple implementation of a memory reference tracing tool
+ * without instrumentation optimization.
+ */
+
+#include <stddef.h> /* for offsetof */
+#include "dr_api.h"
+#include "drmgr.h"
+#include "drutil.h"
+#include "drreg.h"
+#include "utils.h"
+#include "drx.h"
+
+/* We opt to use two buffers -- one to hold only mem_ref_t structs, and another to hold
+ * the raw bytes written. This is done for simplicity, as we will never get a partial
+ * write to the trace buffer (holding mem_ref_t's), which simplifies the handler logic.
+ */
+typedef struct _mem_ref_t {
+    ushort size; /* mem ref size */
+    ushort type; /* instr opcode */
+    app_pc addr; /* mem ref addr */
+} mem_ref_t;
+
+/* Max number of mem_ref a buffer can have. */
+#define MAX_NUM_MEM_REFS 4096
+/* The maximum size of buffer for holding mem_refs. */
+#define MEM_BUF_SIZE (sizeof(mem_ref_t) * MAX_NUM_MEM_REFS)
+/* The maximum size of buffer for holding writes. Writes on average don't get too large,
+ * but we give ourselves some leeway and say chains of consecutive writes are on average
+ * less than 32 bytes each.
+ */
+#define WRT_BUF_SIZE (MAX_NUM_MEM_REFS * 32)
+
+#define MINSERT instrlist_meta_preinsert
+
+/* thread private log file and across-app-inst register */
+typedef struct {
+    file_t     log;
+    FILE      *logf;
+    reg_id_t   reg_addr;
+} per_thread_t;
+
+static client_id_t client_id;
+static int         tls_idx;
+static drx_buf_t  *write_buffer;
+static drx_buf_t  *trace_buffer;
+
+/* Requires that hex_buf be at least as long as 2*memref->size + 1. */
+static char *
+write_hexdump(char *hex_buf, byte *write_base, mem_ref_t *mem_ref)
+{
+    int i;
+    char *hexstring = hex_buf, *needle = hex_buf;
+
+    for (i = mem_ref->size - 1; i >= 0; --i)
+        needle += dr_snprintf(needle, 2*mem_ref->size+1-2*i, "%02x", write_base[i]);
```

Comment:
Safer to use the returned value for the size: so 2*mem_ref->size+1-(needle-hex_buf)